keps/sig-architecture/4330-compatibility-versions/README.md

```diff
@@ -0,0 +1,1175 @@
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+We intend to introduce a "compatibility version" option to Kubernetes control
+plane components to make upgrades safer by increasing the granularity of steps
+available to cluster administrators. The "compatibility version" is distinct from
+the binary version of a Kubernetes component, and can be used to emulate the
+behavior (APIs, features, ...) of a prior Kubernetes release version.
+
+"Compatibility version" is not a bug-for-bug compatibility model. It specifically
+controls which APIs, feature gates, configs and flags are enabled to match
+that of a previous release version.
+
+The [Version skew policy](https://kubernetes.io/releases/version-skew-policy/)
+will apply to compatibility versions, not binary versions, allowing binary
+versions to be skipped when performing skip version upgrades, so long the
+upgrade transitions incrementally through compatibility versions according
+to the supported version skew constraints.
+
+## Motivation
+
+The notion of more granular steps in Kubernetes upgrades is attractive because
+it is more rigorous about how we step through a Kubernetes control-plane
+upgrade, introducing potentially corrupting data (i.e. data only present in N+1
+and not in N) only in later stages of the upgrade process.
+
+For example, upgrading from Kubernetes 1.30 to 1.31 while keeping the compatibility
+version at 1.30 would enable a cluster administrator to validate that the new
+Kubernetes binary version is working as desired before exposing any feature changes
+introduced in 1.31 to cluster users, and without writing and data to storage at
+newer API versions.
+
+This extra step increases the granularity of our upgrade sequence so that 
```

Comment:
> Can I query the component (or api server, or some static tooling) to find out what the feature gates, defaults, etc. may change when I bump the compatibility version?

Yes, I think `kube-apiserver --compatibility-version=1.30 --help` should output `--feature-gate` details with a list of features matching the provided compatibility version.

> Will we have that data or is it built into logic in the code somewhere and so too hard to get?

The plan is for us to book-keep feature gate details (incl. versions where stability level changes) in such a way that we can compute the effective set of feature gates for the provided compatibility version and then use that at runtime.

Reply:
So, presumably you could do `--compatibility-version=1.30`, with binary say, 1.32, but then you could incrementally advance the feature gates until all of them are enabled to be equivalent to 1.31, then bump compat version, then repeat to get to 1.32, then remove compat version. Something like that would be very fine grained. And tiresome. But it should be do-able, right?

Reply:
Yes. This would be possible.

Whenever `-compatibility-version=1.30` is set, the binary will emulate the behavior of 1.30 (for the purposes of APIs, features, storage versions, CEL...) so when compatibility version is set to a particular version, the feature gates can be set to whatever would be allowed for that version.